crates/libm-test/tests/system_libm.rs

```diff
@@ -0,0 +1,135 @@
+        arg_ids: $($arg_ids:ident),*;
+        ret_ty: $ret_ty:ty;
+    ) =>  {};
+    (
+        id: y0f;
+        arg_tys: $($arg_tys:ty),*;
+        arg_ids: $($arg_ids:ident),*;
+        ret_ty: $ret_ty:ty;
+    ) =>  {};
+    (
+        id: y1f;
+        arg_tys: $($arg_tys:ty),*;
+        arg_ids: $($arg_ids:ident),*;
+        ret_ty: $ret_ty:ty;
+    ) =>  {};
+    (
+        id: ynf;
+        arg_tys: $($arg_tys:ty),*;
+        arg_ids: $($arg_ids:ident),*;
+        ret_ty: $ret_ty:ty;
+    ) =>  {};
+    (
+        id: exp10;
+        arg_tys: $($arg_tys:ty),*;
+        arg_ids: $($arg_ids:ident),*;
+        ret_ty: $ret_ty:ty;
+    ) =>  {};
+    (
+        id: exp10f;
+        arg_tys: $($arg_tys:ty),*;
+        arg_ids: $($arg_ids:ident),*;
+        ret_ty: $ret_ty:ty;
+    ) =>  {};
+
+    // Generate random tests for all others:
+    (
+        id: $id:ident;
+        arg_tys: $($arg_tys:ty),*;
+        arg_ids: $($arg_ids:ident),*;
+        ret_ty: $ret_ty:ty;
+    ) => {
+        #[test]
+        #[allow(unused)]
+        fn $id() {
+            use crate::Call;
+            let mut rng = rand::thread_rng();
+            for _ in 0..NTESTS {
+                // Type of the system libm fn:
+                type FnTy
+                    = unsafe extern "C" fn ($($arg_ids: $arg_tys),*) -> $ret_ty;
+
+               extern "C" {
+                    // The system's libm function:
+                    fn $id($($arg_ids: $arg_tys),*) -> $ret_ty;
+                }
+
+                // Generate a tuple of arguments containing random values:
+                let mut args: ( $($arg_tys,)+ )
+                    = ( $(<$arg_tys as Rand>::gen(&mut rng),)+ );
+
+                // Some APIs need their inputs to be "adjusted" (see macro):
+                // correct_input!(fn: $id, input: args);
+                adjust_input!(fn: $id, input: args);
+
+                let result = args.call(libm::$id as FnTy);
+                let expected = args.call($id as FnTy);
+                assert_approx_eq!(
+                    result == expected,
+                    id: $id, arg: args, ulp: ULP_TOL
+                );
+            }
+        }
+    }
+}
+
+libm_analyze::for_each_api!(system_libm);
+
+// We need to be able to generate random numbers for the types involved.
+//
+// Rand does this well, but we want to also test some other specific values.
+trait Rand {
+    fn gen(rng: &mut rand::rngs::ThreadRng) -> Self;
+}
+
+macro_rules! impl_rand {
+    ($id:ident: [$($e:expr),*]) => {
+        impl Rand for $id {
+            fn gen(r: &mut rand::rngs::ThreadRng) -> Self {
+                use rand::{Rng, seq::SliceRandom};
+                // 1/20 probability of picking a non-random value
+                if r.gen_range(0, 20) < 1 {
+                    *[$($e),*].choose(r).unwrap()
+                } else {
+                    r.gen::<$id>()
+                }
+            }
+        }
+    }
+}
+
+// Some interesting random values
```

Comment:
We should properly think about adding positive and negative NAN, MIN_POSITIVE, MAX and, subnorm which are `1.401298464324817070923730e-45f32` and `4.9406564584124654e-324f64`

Reply:
I tried to keep the tests testing exactly the same thing as in the old `build.rs`, to keep the test from breaking on CI due to too many new issues - there were already enough simpler to fix issues being raised by libm-analyze. I think we should move this to do something cleverer here as we fix the tests.

Reply:
I've added some of these values, haven't added the subnormal yet, I thought `MIN_POSITIVE` and `-MIN_POSITIVE` where the two subnormal values. Is that incorrect ?